src/pykeen/nn/node_piece/representations.py

```diff
@@ -0,0 +1,301 @@
+
+"""Representation modules for NodePiece."""
+
+import logging
+from typing import Callable, Optional, Sequence, Union
+
+import torch
+from class_resolver import HintOrType, OneOrManyHintOrType, OneOrManyOptionalKwargs, OptionalKwargs
+from class_resolver.contrib.torch import aggregation_resolver
+
+from .tokenization import Tokenizer, tokenizer_resolver
+from ..representation import Representation
+from ...triples import CoreTriplesFactory
+from ...typing import MappedTriples, OneOrSequence
+from ...utils import broadcast_upgrade_to_sequences
+
+__all__ = [
+    "TokenizationRepresentation",
+    "NodePieceRepresentation",
+]
+
+logger = logging.getLogger(__name__)
+
+
+class TokenizationRepresentation(Representation):
+    """A module holding the result of tokenization."""
+
+    #: the token ID of the padding token
+    vocabulary_size: int
+
+    #: the token representations
+    vocabulary: Representation
+
+    #: the assigned tokens for each entity
+    assignment: torch.LongTensor
+
+    def __init__(
+        self,
+        assignment: torch.LongTensor,
+        token_representation: HintOrType[Representation] = None,
+        token_representation_kwargs: OptionalKwargs = None,
+        **kwargs,
+    ) -> None:
+        """
+        Initialize the tokenization.
+
+        :param assignment: shape: `(n, num_chosen_tokens)`
+            the token assignment.
+        :param token_representation: shape: `(num_total_tokens, *shape)`
+            the token representations
+        :param token_representation_kwargs:
+            additional keyword-based parameters
+        :param kwargs:
+            additional keyword-based parameters passed to super.__init__
+        :raises ValueError: if there's a mismatch between the representation size
+            and the vocabulary size
+        """
+        # needs to be lazily imported to avoid cyclic imports
+        from .. import representation_resolver
+
+        # fill padding (nn.Embedding cannot deal with negative indices)
+        padding = assignment < 0
+        # sometimes, assignment.max() does not cover all relations (eg, inductive inference graphs
+        # contain a subset of training relations) - for that, the padding index is the last index of the Representation
+        self.vocabulary_size = (
+            token_representation.max_id
+            if isinstance(token_representation, Representation)
+            else assignment.max().item() + 2  # exclusive (+1) and including padding (+1)
+        )
+
+        assignment[padding] = self.vocabulary_size - 1  # = assignment.max().item() + 1
+        max_id, num_chosen_tokens = assignment.shape
+
+        # resolve token representation
+        token_representation = representation_resolver.make(
+            token_representation,
+            token_representation_kwargs,
+            max_id=self.vocabulary_size,
+            # FIXME @mberr not sure why there's neither ``embedding_dim`` nor ``shape`` here, but that's the issue
```

Comment:
@mberr not sure why this is causing an error now, but it looks like we're missing some things that need to be passed to the representation resolver

Reply:
fc5b040